File: Python/ScientificCalculator/functions/binomial.py
"""
binomial.py

Binomial expansions up to power 10.

Supports:

expand_plus(n)
expand_minus(n)
pascal_row(n)
"""

PASCAL = {
    0: [1],
    1: [1, 1],
    2: [1, 2, 1],
    3: [1, 3, 3, 1],
    4: [1, 4, 6, 4, 1],
    5: [1, 5, 10, 10, 5, 1],
    6: [1, 6, 15, 20, 15, 6, 1],
    7: [1, 7, 21, 35, 35, 21, 7, 1],
    8: [1, 8, 28, 56, 70, 56, 28, 8, 1],
    9: [1, 9, 36, 84, 126, 126, 84, 36, 9, 1],
    10: [1, 10, 45, 120, 210, 252, 210, 120, 45, 10, 1],
}


def pascal_row(power: int):
    """
    Returns Pascal row.

    Example:
        pascal_row(5)
    """

    if power not in PASCAL:
        raise ValueError("Supported powers: 0-10")

    return PASCAL[power]


def _power(variable: str, exponent: int):

    if exponent == 0:
        return ""

    if exponent == 1:
        return variable

    return f"{variable}^{exponent}"


def expand_plus(power: int):
    """
    Returns expansion of (a+b)^power
    """

    if power not in PASCAL:
        raise ValueError("Supported powers: 0-10")

    coeffs = PASCAL[power]

    terms = []

    for i, coeff in enumerate(coeffs):

        a_exp = power - i
        b_exp = i

        term = ""

        if coeff != 1:
            term += str(coeff)

        term += _power("a", a_exp)
        term += _power("b", b_exp)

        if term == "":
            term = "1"

        terms.append(term)

    return " + ".join(terms)


def expand_minus(power: int):
    """
    Returns expansion of (a-b)^power
    """

    if power not in PASCAL:
        raise ValueError("Supported powers: 0-10")

    coeffs = PASCAL[power]

    expression = ""

    for i, coeff in enumerate(coeffs):

        a_exp = power - i
        b_exp = i

        term = ""

        if coeff != 1:
            term += str(coeff)

        term += _power("a", a_exp)
        term += _power("b", b_exp)

        if term == "":
            term = "1"

        if i == 0:
            expression += term
        else:
            if i % 2 == 1:
                expression += " - " + term
            else:
                expression += " + " + term

    return expression


if __name__ == "__main__":

    print("(a+b)^5")
    print(expand_plus(5))
    print()

    print("(a-b)^5")
    print(expand_minus(5))
    print()

    print("Pascal Row 10")
    print(pascal_row(10))

File: Java/ScientificCalculator/src/functions/Binomial.java
package functions;


/*
 * Binomial
 *
 * Expands:
 *
 * (a + b)^n
 * (a - b)^n
 *
 */

public class Binomial {


    // ==========================
    // Plus Expansion
    //
    // (a+b)^n
    // ==========================

    public static String expandPlus(
            String a,
            String b,
            int n
    ) {

        return expand(
                a,
                b,
                n,
                true
        );

    }



    // ==========================
    // Minus Expansion
    //
    // (a-b)^n
    // ==========================

    public static String expandMinus(
            String a,
            String b,
            int n
    ) {

        return expand(
                a,
                b,
                n,
                false
        );

    }



    // ==========================
    // Pascal Row
    // ==========================

    public static String pascalRow(
            int n
    ) {


        StringBuilder result =
                new StringBuilder();


        for (
                int k = 0;
                k <= n;
                k++
        ) {


            result.append(
                    combination(n,k)
            );


            if (k != n) {

                result.append(", ");

            }

        }


        return result.toString();

    }



    // ==========================
    // Expansion Engine
    // ==========================

    private static String expand(
            String a,
            String b,
            int n,
            boolean plus
    ) {


        StringBuilder result =
                new StringBuilder();


        for (
                int i = 0;
                i <= n;
                i++
        ) {


            int coefficient =
                    combination(n,i);



            int powerA =
                    n-i;


            int powerB =
                    i;



            if (i > 0) {


                if (plus || i % 2 == 0) {

                    result.append(" + ");

                }

                else {

                    result.append(" - ");

                }

            }



            if (coefficient != 1) {

                result.append(coefficient);

            }



            if (powerA > 0) {

                result.append(a);

                if (powerA > 1) {

                    result.append("^")
                          .append(powerA);

                }

            }



            if (powerB > 0) {


                result.append(b);


                if (powerB > 1) {

                    result.append("^")
                          .append(powerB);

                }

            }


        }


        return result.toString();

    }



    // ==========================
    // Combination
    // nCr
    // ==========================

    private static int combination(
            int n,
            int r
    ) {


        return factorial(n)
                /
                (
                    factorial(r)
                    *
                    factorial(n-r)
                );

    }



    private static int factorial(
            int n
    ) {

        int result = 1;


        for (
                int i = 2;
                i <= n;
                i++
        ) {

            result *= i;

        }


        return result;

    }

}
